File: _1_Hashmap/Session1.java
package _1_Hashmap;

import java.util.Scanner;

public class Session1 {

    // static boolean distanceLessThanK(int[] arr, int n, int k) {
    //     //  Brute Force: Implement the logic to find pairs with distance less than k
    //     int count = 0;
    //     for (int i = 0; i < n; i++) {
    //         for (int j = i + 1; j <= n && j<= i+k; j++) {
    //             if(arr[j] == arr[i]){
    //                 System.out.println("Pair Found: " + i + ", " + j );
    //                 return true;
    //             }
    //         }
    //     }
    //     return false;
    // }

    static boolean distanceLessThanK(int[] arr, int n, int k) {
        // Optimal Approach: Implement the logic to find pairs with distance less than k
        java.util.HashMap<Integer, Integer> map = new java.util.HashMap<>();
        for (int i = 0; i < n; i++) {
            if (map.containsKey(arr[i])) {
                int prevIndex = map.get(arr[i]);
                if (i - prevIndex <= k) {
                    System.out.println("Pair Found: " + prevIndex + ", " + i);
                    return true;
                }
            }
            map.put(arr[i], i);
        }
        return false;
    }
    

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        int k = sc.nextInt();
        // Your code here
        distanceLessThanK(arr, n, k);
    }
}

File: _1_Hashmap/Session1.py
def distance_less_than_k_brute(arr, n, k):
    # Brute Force: Check all pairs within distance k
    for i in range(n):
        for j in range(i + 1, min(n, i + k + 1)):
            if arr[i] == arr[j]:
                print(f"Pair Found: {i}, {j}")
                return True
    return False


def distance_less_than_k_optimal(arr, n, k):
    # Optimal Approach: Using HashMap (dictionary)
    index_map = {}

    for i in range(n):
        if arr[i] in index_map:
            prev_index = index_map[arr[i]]
            if i - prev_index <= k:
                print(f"Pair Found: {prev_index}, {i}")
                return True
        index_map[arr[i]] = i

    return False


def main():
    n = int(input("Enter n: "))
    arr = list(map(int, input("Enter array elements: ").split()))
    k = int(input("Enter k: "))

    print("\n--- Brute Force ---")
    found_brute = distance_less_than_k_brute(arr, n, k)
    if not found_brute:
        print("No pair found using Brute Force")

    print("\n--- Optimal Approach ---")
    found_optimal = distance_less_than_k_optimal(arr, n, k)
    if not found_optimal:
        print("No pair found using Optimal Approach")


if __name__ == "__main__":
    main()
